recipes/dev/7B_full_early_exit.yaml

```diff
@@ -0,0 +1,140 @@
+#
+# To launch on 4 devices, run the following command from root:
+#   tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml
+#
+# You can add specific overrides through the command line. For example
+# to override the checkpointer directory while launching training
+# you can run:
+#   tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml checkpointer.checkpoint_dir=<YOUR_CHECKPOINT_DIR>
+#
+# To reproduce experiments of various papers that use early exit loss and/or layer dropout:
+# - LayerSkip (https://arxiv.org/abs/2404.16710) on TOPv2:
+#   tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml early_exit_loss.scale=1.0, early_exit_loss.curriculum=gradual early_exit_loss.scale_type=l layer_dropout.prob=0.2 layer_dropout.scale=exp
+#
+# - LITE (https://arxiv.org/abs/2310.18581):
+#   tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml layer_dropout=null early_exit_loss.layers=8,12,16,20,24,28 early_exit_loss.scale_type=one
+#
+# - LayerDrop (https://arxiv.org/abs/1909.11556):
+#   tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml early_exit_loss=null layer_dropout.prob=0.2 layer_dropout.layers=::2
+#
+# - Progressive Layer Dropping (https://arxiv.org/abs/2010.13369) (The paper also implements a curriculum for layer drop probability which is not yet implemented.):
+#   tune run --nnodes 1 --nproc_per_node 4 dev/early_exit_finetune_distributed --config recipes/dev/7B_full_early_exit.yaml early_exit_loss=null layer_dropout.prob=0.5 layer_dropout.scale=exp
+#
+# This config works best for distributed training, hence when the model is being fine-tuned on 2+ GPUs.
+#
+
+
+# Tokenizer
+tokenizer:
+  _component_: torchtune.models.llama2.llama2_tokenizer
+  path: /tmp/Llama-2-7b-hf/tokenizer.model
+  max_seq_len: null
+
+# Dataset
+# dataset:
+#   _component_: torchtune.datasets.alpaca_dataset
+#   packed: False  # True increases speed
+dataset:
+  _component_: torchtune.datasets.instruct_dataset
+  source: WillHeld/top_v2
+  split: train
+  column_map:
+    input: utterance
+    output: semantic_parse
+
+seed: null
+shuffle: True
+
+# Model Arguments
+model:
+  _component_: torchtune.models.llama2.llama2_7b
+
+checkpointer:
+  _component_: torchtune.training.FullModelHFCheckpointer
+  checkpoint_dir: /tmp/Llama-2-7b-hf
+  checkpoint_files: [
+    pytorch_model-00001-of-00002.bin,
+    pytorch_model-00002-of-00002.bin
+  ]
+  recipe_checkpoint: null
+  output_dir: /tmp/Llama-2-7b-hf
+  model_type: LLAMA2
+resume_from_checkpoint: False
+
+# Fine-tuning arguments
+batch_size: 2
+epochs: 1
+optimizer:
+  _component_: torch.optim.AdamW
+  fused: True
+  lr: 2e-5
+loss:
+  _component_: torch.nn.CrossEntropyLoss
```

Comment:
In principal, early exit loss should work with any loss function including chunked cross entropy. However, the current implementation doesn't support it.
The current implementation is parallelized: it stacks the tensor outputs of various layers into one big tensor, applies the language model head on that tensor, and then applies the loss function once. 
For chunked cross entropy, applying the language model head creates a list rather than a tensor and I found that difficult to tensorize. So in this PR I didn't implement it (I added an assertion statement to throw an error if that's the case).

So we have 2 options:
Option A We can have a simple albeit less efficient implementation that doesn't parallelize: just loop over the outputs of different layers and apply the language model head. This way, we can guarantee supporting any type of loss. We can have a condition or heurestic to guide us whether we can apply early exit losses across layers in parallel or in sequence.
Option B: not sure if there is a way to parallelize some sort of lambda function across various layers. Does functorch help with that
Any other options?

Not sure if you would like to implement any of the options above in this PR or in a future PR. For this PR, I will need to re-run the training experiments to verify we get speedups during inference.

Reply:
Thanks for the additional information. It sounds to me like either of these two options would involve some kind of extension of the early_exit_loss as you've defined it, so I think leaving it for a follow-up makes sense (especially given that you already assert to prevent accidentally using with our `CEWithChunkedOutputLoss`).